pkg/controller/acmechallenges/sync.go

```diff
@@ -266,39 +252,29 @@ func handleError(ch *cmacme.Challenge, err error) error {
 // CleanUp if the resource is in a 'processing' state.
 func (c *controller) handleFinalizer(ctx context.Context, ch *cmacme.Challenge) (err error) {
 	log := logf.FromContext(ctx, "finalizer")
-	if len(ch.Finalizers) == 0 {
-		return nil
-	}
-	if otherFinalizerPresent(ch) {
-		log.V(logf.DebugLevel).Info("waiting to run challenge finalization...")
-		return nil
-	}
-
-	defer func() {
-		// call Update to remove the metadata.finalizers entry
-		ch.Finalizers = slices.DeleteFunc(ch.Finalizers, func(finalizer string) bool {
-			return finalizer == cmacme.ACMELegacyFinalizer || finalizer == cmacme.ACMEDomainQualifiedFinalizer
-		})
-	}()
-
-	if !ch.Status.Processing {
-		return nil
+	if !hasFinalizer(ch) {
+		return nil // nothing to do
 	}
 
 	solver, err := c.solverFor(ch.Spec.Type)
 	if err != nil {
 		log.Error(err, "error getting solver for challenge")
-		return nil
```

Comment:
question: Previously we swallowed the error and returned `nil`. Same in the `if err := solver.CleanUp(ctx, ch); err != nil` block below.

What's the motivation for the change to now start returning the error here (and below)? What effect will it have?

Reply:
There are two situations in which a cleanup is performed:
- `acme.IsFinalState(ch.Status.State)`
- `!ch.DeletionTimestamp.IsZero()`

The old code swallowed errors when `!ch.DeletionTimestamp.IsZero()` but returned errors when `acme.IsFinalState(ch.Status.State)`. So this change will only affect the first case.
In that case, the deletion of the challenge resource will be blocked until the `handleFinalizer` is successful. An error will be returned if something went wrong and the `CleanUp` function will be retried.
The previous behavior for the first case was that the `CleanUp` function was called once and the error was only logged and the finalizer deleted unconditionally.

Reply:
So, the new behavior will block deletion until the cleanup succeeds (which was what is described here: https://github.com/cert-manager/cert-manager/issues/3640).

Reply:
🤔 I just found there is a counter-issue: https://github.com/cert-manager/cert-manager/issues/7234
Challenges that are stuck in their cleanup phase: cleanup is failing and the resource is not being deleted...